File: leetcode/704-binary-search/main.go
/*
Problem: Binary Search
Link: https://leetcode.com/problems/binary-search/
*/
package main

func binarySearch(nums []int, target int) int {
	start := 0
	end := len(nums) - 1

	for start <= end {
		i := (start + end) / 2

		if nums[i] > target {
			end = i - 1
		} else if nums[i] < target {
			start = i + 1
		} else {
			return i
		}
	}

	return -1
}

File: leetcode/704-binary-search/Main.kt
package main

fun main() {
  val solution = Solution()
  println(solution.search(nums = intArrayOf(-1, 0, 3, 5, 9, 12), target = 9))
  // 4
}

class Solution {
  fun search(nums: IntArray, target: Int): Int {
    var start = 0
    var end = nums.size - 1

    while (start <= end) {
      val mid = (start + end) / 2

      if (nums[mid] > target) {
        end = mid - 1
      } else if (nums[mid] < target) {
        start = mid + 1
      } else {
        return mid
      }
    }

    return -1
  }
}